keps/sig-storage/0661-statefulset-volume-resize/README.md

```diff
@@ -0,0 +1,808 @@
+process of resizing StatefulSet volumes.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+Allow users to resize the volumes associated with `StatefulSet` by modifying the size of
+PVC in the `volumeClaimTemplate`.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+This proposal does not try to address the following scenarios but relies on the
+underlying PVC resize logic to take the appropriate action or indicate error as needed.
+- Shrinking of volume is not allowed.
+- Few environments require Pod restarts in order
+to get access to the resized volume.
+- Modifying the storage class of the volumeClaimTemplate
+- Resizing a PVC associated with a statefulset directly by modifying the PVC spec will not affect the 
+  Statefulset `volumeClaimTemplates`. 
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+Before we proceed to the proposal it is imperative to layout the various components involved
+in creating and managing the persistent volumes associated with `StatefulSet`. 
+
+- StatefulSet controller creates the PVCs from `volumeClaimTemplates` and associates them with the pods
+spawned from it. Reconciliation logic runs for every change made to the StatefulSet
+`Spec` and `Status` reflects the progress towards making the real world state equal to the
+`Spec`.
+- The admission controller in the API server is responsible for restricting the fields
+which can be edited in a StatefulSet specification.
+- The RBAC specification in the controller policy is responsible for adding the required 
+RBAC permission for StatefulSet.
+- The StatefulSet controller also supports revision control.
+
+The following changes are proposed to achieve the goal:
+1. Make the `volumeClaimTemplates` storage size editable by modification to the api server
+   validation. Some validation checks will depend on the design decisions made in [KEP 1790](../1790-recover-resize-failure/README.md) to prevent user from shrinking the volume.
+2. Add reconciliation of the associated PVC's size and the individual `volumeClaimTemplates` size into the 
+   StatefulSet controller reconciliation.
+3. Add PVC `update` privilege to the StatefulSet controller's clusterrole RBAC.
+
+Once the above changes are made, modification to any of the `volumeClaimTemplates` storage
+size by user will update the underlying PVC via the reconciliation performed by (2) above.
+The existing PVC resize workflow kicks in at this point and attempts to resize the volume.
```

Comment:
We should specify how we mean to apply those changes to `volumeClaimTemplate`. Such as:

1. Are we going to perform changes to PVC one-by-one and wait for resize to be finished on PVCs? Is there going to be an status or something else added to statefulset to indicate that an expansion is happening?
2. What about volume types that only allow offline expansion? Is statefulset controller going to delete pods and expand the corresponding volume and bring them back?
3.